File: src/tui/edit-group.ts
import * as p from "@clack/prompts";
import path from "node:path";
import {
  addReposToGroup,
  removeReposFromGroup,
  renameGroup,
  updateGroup,
} from "./mutations.js";
import { detectSiblings } from "./siblings.js";
import {
  formatCommaList,
  formatLaunchCommand,
  parseCommaList,
  parseLaunchCommand,
} from "./parse.js";
import {
  resolveRepoPath,
  validateGroupName,
  validateWorktreePattern,
} from "./prompts.js";
import type { Config } from "../config/schema.js";

export async function editGroupWizard(config: Config, cwd: string): Promise<Config | null> {
  if (config.groups.length === 0) {
    p.log.info("No groups to edit.");
    return null;
  }

  const pick = await p.select({
    message: "Which group?",
    options: config.groups.map((g) => ({
      value: g.name,
      label: g.name,
      hint: `${g.repos.length} repos`,
    })),
  });
  if (p.isCancel(pick)) return null;
  const groupName = pick as string;

  let current = config;
  let activeName = groupName;

  while (true) {
    const field = await p.select({
      message: `Editing "${activeName}" — which field?`,
      options: [
        { value: "name", label: "Rename group" },
        { value: "repos", label: "Edit repos" },
        { value: "symlinks", label: "Override symlinkDirectories" },
        { value: "pattern", label: "Override worktreePathPattern" },
        { value: "launch", label: "Override launchCommand" },
        { value: "back", label: "← Back to main menu" },
      ],
    });
    if (p.isCancel(field) || field === "back") return current === config ? null : current;

    switch (field) {
      case "name": {
        const group = current.groups.find((g) => g.name === activeName)!;
        const others = current.groups.map((g) => g.name).filter((n) => n !== group.name);
        const newName = await p.text({
          message: "New name",
          initialValue: group.name,
          validate: (v) => validateGroupName(others, v),
        });
        if (p.isCancel(newName)) break;
        current = renameGroup(current, activeName, (newName as string).trim());
        activeName = (newName as string).trim();
        break;
      }
      case "repos": {
        const next = await editRepos(current, activeName, cwd);
        if (next !== null) current = next;
        break;
      }
      case "symlinks": {
        const group = current.groups.find((g) => g.name === activeName)!;
        const input = await p.text({
          message: "symlinkDirectories (comma-separated, blank to unset override)",
          placeholder: formatCommaList(current.symlinkDirectories),
          initialValue: group.symlinkDirectories ? formatCommaList(group.symlinkDirectories) : "",
        });
        if (p.isCancel(input)) break;
        const patch =
          (input as string).trim().length > 0
            ? { symlinkDirectories: parseCommaList(input as string) }
            : { symlinkDirectories: undefined };
        current = updateGroup(current, activeName, patch);
        break;
      }
      case "pattern": {
        const group = current.groups.find((g) => g.name === activeName)!;
        const input = await p.text({
          message: "worktreePathPattern (blank to unset override)",
          placeholder: current.worktreePathPattern,
          initialValue: group.worktreePathPattern ?? "",
          validate: validateWorktreePattern,
        });
        if (p.isCancel(input)) break;
        const patch =
          (input as string).trim().length > 0
            ? { worktreePathPattern: (input as string).trim() }
            : { worktreePathPattern: undefined };
        current = updateGroup(current, activeName, patch);
        break;
      }
      case "launch": {
        const group = current.groups.find((g) => g.name === activeName)!;
        const input = await p.text({
          message: "launchCommand (space-separated, blank to unset override)",
          placeholder: formatLaunchCommand(current.launchCommand),
          initialValue: group.launchCommand ? formatLaunchCommand(group.launchCommand) : "",
        });
        if (p.isCancel(input)) break;
        const patch =
          (input as string).trim().length > 0
            ? { launchCommand: parseLaunchCommand(input as string) }
            : { launchCommand: undefined };
        current = updateGroup(current, activeName, patch);
        break;
      }
    }
  }
}

async function editRepos(config: Config, groupName: string, cwd: string): Promise<Config | null> {
  let current = config;
  while (true) {
    const group = current.groups.find((g) => g.name === groupName)!;
    const action = await p.select({
      message: `Repos (${group.repos.length}): ${group.repos.map((r) => path.basename(r)).join(", ")}`,
      options: [
        { value: "add", label: "Add a repo" },
        {
          value: "remove",
          label: "Remove repos",
          hint: group.repos.length <= 2 ? "need at least 2" : "",
        },
        { value: "done", label: "← Done" },
      ],
    });
    if (p.isCancel(action) || action === "done") return current;

    if (action === "add") {
      const toAdd = await collectReposToAdd(cwd, group.repos);
      if (toAdd === null) return current;
      if (toAdd.length > 0) current = addReposToGroup(current, groupName, toAdd);
    } else if (action === "remove") {
      if (group.repos.length <= 2) {
        p.log.error("Cannot remove — a group must have at least 2 repos.");
        continue;
      }
      const toRemove = await p.multiselect({
        message: "Select repos to remove",
        options: group.repos.map((r) => ({ value: r, label: path.basename(r), hint: r })),
        required: false,
      });
      if (p.isCancel(toRemove)) return current;
      if ((toRemove as string[]).length > 0) {
        try {
          current = removeReposFromGroup(current, groupName, toRemove as string[]);
        } catch (err) {
          p.log.error((err as Error).message);
        }
      }
    }
  }
}

async function collectReposToAdd(
  cwd: string,
  alreadyIn: readonly string[],
): Promise<string[] | null> {
  const siblings = (await detectSiblings(cwd)).filter((s) => !alreadyIn.includes(s));

  const added: string[] = [];
  if (siblings.length > 0) {
    const picks = await p.multiselect({
      message: "Which sibling repos to add?",
      options: siblings.map((s) => ({ value: s, label: path.basename(s), hint: s })),
      required: false,
    });
    if (p.isCancel(picks)) return null;
    added.push(...(picks as string[]));
  }

  const wantManual = await p.confirm({
    message: "Add manual paths?",
    initialValue: added.length === 0,
  });
  if (p.isCancel(wantManual)) return null;

  if (wantManual) {
    while (true) {
      const raw = await p.text({
        message: "Repo path (blank to finish)",
        placeholder: "~/code/another-repo",
      });
      if (p.isCancel(raw)) return null;
      if ((raw as string).trim().length === 0) break;

      const result = await resolveRepoPath(raw as string);
      if (!result.ok) {
        p.log.error(result.error ?? "Invalid path");
        continue;
      }
      if (alreadyIn.includes(result.resolved!) || added.includes(result.resolved!)) {
        p.log.warn("Already added.");
        continue;
      }
      added.push(result.resolved!);
    }
  }

  return added;
}

File: internal/tui/editgroup.go
package tui

import (
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"github.com/charmbracelet/huh"

	"github.com/OysterD3/wtmux/internal/config"
)

// EditGroupWizard runs the edit-a-group flow. Returns nil on cancel without
// changes; otherwise the mutated config.
func EditGroupWizard(cfg *config.Config, cwd string) (*config.Config, error) {
	if len(cfg.Groups) == 0 {
		fmt.Fprintln(os.Stderr, "[wtmux] No groups to edit.")
		return nil, nil
	}

	groupOpts := make([]huh.Option[string], 0, len(cfg.Groups))
	for _, g := range cfg.Groups {
		groupOpts = append(groupOpts, huh.NewOption(fmt.Sprintf("%s  (%d repos)", g.Name, len(g.Repos)), g.Name))
	}
	var picked string
	if err := huh.NewSelect[string]().
		Title("Which group?").
		Options(groupOpts...).
		Value(&picked).
		Run(); err != nil {
		if errors.Is(err, huh.ErrUserAborted) {
			return nil, nil
		}
		return nil, err
	}

	current := cfg
	activeName := picked
	mutated := false

	for {
		var field string
		if err := huh.NewSelect[string]().
			Title(fmt.Sprintf("Editing %q — which field?", activeName)).
			Options(
				huh.NewOption("Rename group", "name"),
				huh.NewOption("Edit repos", "repos"),
				huh.NewOption("Override symlinkDirectories", "symlinks"),
				huh.NewOption("Override worktreePathPattern", "pattern"),
				huh.NewOption("Override launchCommand", "launch"),
				huh.NewOption("Override agent (sibling-injection strategy)", "agent"),
				huh.NewOption("← Back to main menu", "back"),
			).
			Value(&field).
			Run(); err != nil {
			if errors.Is(err, huh.ErrUserAborted) || field == "back" {
				if !mutated {
					return nil, nil
				}
				return current, nil
			}
			return nil, err
		}
		if field == "back" {
			if !mutated {
				return nil, nil
			}
			return current, nil
		}

		idx := findGroupIndex(current, activeName)
		if idx < 0 {
			return current, nil
		}
		group := current.Groups[idx]

		switch field {
		case "name":
			others := make([]string, 0, len(current.Groups)-1)
			for _, g := range current.Groups {
				if g.Name != group.Name {
					others = append(others, g.Name)
				}
			}
			newName := group.Name
			if err := huh.NewInput().
				Title("New name").
				Value(&newName).
				Validate(func(v string) error { return ValidateGroupName(others, v) }).
				Run(); err != nil {
				if errors.Is(err, huh.ErrUserAborted) {
					continue
				}
				return nil, err
			}
			renamed := strings.TrimSpace(newName)
			next, err := config.RenameGroup(current, activeName, renamed)
			if err != nil {
				fmt.Fprintf(os.Stderr, "[wtmux] %s\n", err)
				continue
			}
			current = next
			activeName = renamed
			mutated = true
		case "repos":
			next, ok, err := editRepos(current, activeName, cwd)
			if err != nil {
				return nil, err
			}
			if ok {
				current = next
				mutated = true
			}
		case "symlinks":
			input := strings.Join(group.SymlinkDirectories, ", ")
			if err := huh.NewInput().
				Title("symlinkDirectories (comma-separated, blank to unset override)").
				Placeholder(strings.Join(current.SymlinkDirectories, ", ")).
				Value(&input).
				Run(); err != nil {
				if errors.Is(err, huh.ErrUserAborted) {
					continue
				}
				return nil, err
			}
			updated := group
			if strings.TrimSpace(input) != "" {
				updated.SymlinkDirectories = ParseCommaList(input)
			} else {
				updated.SymlinkDirectories = nil
			}
			next, err := config.UpdateGroup(current, activeName, updated)
			if err != nil {
				fmt.Fprintf(os.Stderr, "[wtmux] %s\n", err)
				continue
			}
			current = next
			mutated = true
		case "pattern":
			input := group.WorktreePathPattern
			if err := huh.NewInput().
				Title("worktreePathPattern (blank to unset override)").
				Placeholder(current.WorktreePathPattern).
				Value(&input).
				Validate(ValidateWorktreePattern).
				Run(); err != nil {
				if errors.Is(err, huh.ErrUserAborted) {
					continue
				}
				return nil, err
			}
			updated := group
			if strings.TrimSpace(input) != "" {
				updated.WorktreePathPattern = strings.TrimSpace(input)
			} else {
				updated.WorktreePathPattern = ""
			}
			next, err := config.UpdateGroup(current, activeName, updated)
			if err != nil {
				fmt.Fprintf(os.Stderr, "[wtmux] %s\n", err)
				continue
			}
			current = next
			mutated = true
		case "launch":
			input := strings.Join(group.LaunchCommand, " ")
			if err := huh.NewInput().
				Title("launchCommand (space-separated, blank to unset override)").
				Placeholder(strings.Join(current.LaunchCommand, " ")).
				Value(&input).
				Run(); err != nil {
				if errors.Is(err, huh.ErrUserAborted) {
					continue
				}
				return nil, err
			}
			updated := group
			if strings.TrimSpace(input) != "" {
				updated.LaunchCommand = ParseLaunchCommand(input)
			} else {
				updated.LaunchCommand = nil
			}
			next, err := config.UpdateGroup(current, activeName, updated)
			if err != nil {
				fmt.Fprintf(os.Stderr, "[wtmux] %s\n", err)
				continue
			}
			current = next
			mutated = true
		case "agent":
			result, err := AgentSelect(AgentSelectInput{
				Title:             "Agent (sibling-injection strategy)",
				CurrentAgent:      group.Agent,
				CurrentAddDirArgs: group.AddDirArgs,
			})
			if err != nil {
				return nil, err
			}
			if result.Cancelled {
				continue
			}
			updated := group
			updated.Agent = result.Agent
			updated.AddDirArgs = result.AddDirArgs
			next, err := config.UpdateGroup(current, activeName, updated)
			if err != nil {
				fmt.Fprintf(os.Stderr, "[wtmux] %s\n", err)
				continue
			}
			current = next
			mutated = true
		}
	}
}

func editRepos(cfg *config.Config, groupName, cwd string) (*config.Config, bool, error) {
	current := cfg
	mutated := false
	for {
		idx := findGroupIndex(current, groupName)
		if idx < 0 {
			return current, mutated, nil
		}
		group := current.Groups[idx]

		baseNames := make([]string, 0, len(group.Repos))
		for _, r := range group.Repos {
			baseNames = append(baseNames, filepath.Base(r))
		}
		var action string
		removeHint := ""
		if len(group.Repos) <= 2 {
			removeHint = " (need at least 2)"
		}
		if err := huh.NewSelect[string]().
			Title(fmt.Sprintf("Repos (%d): %s", len(group.Repos), strings.Join(baseNames, ", "))).
			Options(
				huh.NewOption("Add a repo", "add"),
				huh.NewOption("Remove repos"+removeHint, "remove"),
				huh.NewOption("← Done", "done"),
			).
			Value(&action).
			Run(); err != nil {
			if errors.Is(err, huh.ErrUserAborted) {
				return current, mutated, nil
			}
			return nil, false, err
		}

		switch action {
		case "done":
			return current, mutated, nil
		case "add":
			toAdd, cancelled, err := collectReposToAdd(cwd, group.Repos)
			if err != nil {
				return nil, false, err
			}
			if cancelled || len(toAdd) == 0 {
				continue
			}
			next, err := config.AddReposToGroup(current, groupName, toAdd)
			if err != nil {
				fmt.Fprintf(os.Stderr, "[wtmux] %s\n", err)
				continue
			}
			current = next
			mutated = true
		case "remove":
			if len(group.Repos) <= 2 {
				fmt.Fprintln(os.Stderr, "[wtmux] Cannot remove — a group must have at least 2 repos.")
				continue
			}
			opts := make([]huh.Option[string], 0, len(group.Repos))
			for _, r := range group.Repos {
				opts = append(opts, huh.NewOption(filepath.Base(r)+"  "+r, r))
			}
			var picked []string
			if err := huh.NewMultiSelect[string]().
				Title("Select repos to remove").
				Options(opts...).
				Value(&picked).
				Run(); err != nil {
				if errors.Is(err, huh.ErrUserAborted) {
					continue
				}
				return nil, false, err
			}
			if len(picked) == 0 {
				continue
			}
			next, err := config.RemoveReposFromGroup(current, groupName, picked)
			if err != nil {
				fmt.Fprintf(os.Stderr, "[wtmux] %s\n", err)
				continue
			}
			current = next
			mutated = true
		}
	}
}

func collectReposToAdd(cwd string, alreadyIn []string) ([]string, bool, error) {
	siblings := DetectSiblings(cwd)
	excluded := map[string]bool{}
	for _, r := range alreadyIn {
		excluded[r] = true
	}
	filtered := make([]string, 0, len(siblings))
	for _, s := range siblings {
		if !excluded[s] {
			filtered = append(filtered, s)
		}
	}

	var added []string
	if len(filtered) > 0 {
		opts := make([]huh.Option[string], 0, len(filtered))
		for _, s := range filtered {
			opts = append(opts, huh.NewOption(filepath.Base(s)+"  "+s, s))
		}
		if err := huh.NewMultiSelect[string]().
			Title("Which sibling repos to add?").
			Options(opts...).
			Value(&added).
			Run(); err != nil {
			if errors.Is(err, huh.ErrUserAborted) {
				return nil, true, nil
			}
			return nil, false, err
		}
	}

	wantManual := len(added) == 0
	if err := huh.NewConfirm().
		Title("Add manual paths?").
		Value(&wantManual).
		Run(); err != nil {
		if errors.Is(err, huh.ErrUserAborted) {
			return nil, true, nil
		}
		return nil, false, err
	}

	if wantManual {
		for {
			var raw string
			if err := huh.NewInput().
				Title("Repo path (blank to finish)").
				Placeholder("~/code/another-repo").
				Value(&raw).
				Run(); err != nil {
				if errors.Is(err, huh.ErrUserAborted) {
					return nil, true, nil
				}
				return nil, false, err
			}
			if strings.TrimSpace(raw) == "" {
				break
			}
			result := ResolveRepoPath(raw)
			if !result.OK {
				fmt.Fprintf(os.Stderr, "[wtmux] %s\n", result.Reason)
				continue
			}
			if containsString(alreadyIn, result.Resolved) || containsString(added, result.Resolved) {
				fmt.Fprintln(os.Stderr, "[wtmux] Already added.")
				continue
			}
			added = append(added, result.Resolved)
		}
	}

	return added, false, nil
}
